library/psa_crypto_slot_management.c

```diff
@@ -276,6 +277,66 @@ static psa_status_t psa_load_persistent_key_into_slot( psa_key_slot_t *slot )
 }
 #endif /* MBEDTLS_PSA_CRYPTO_STORAGE_C */
 
+#if defined(MBEDTLS_PSA_CRYPTO_BUILTIN_KEYS)
+
+static psa_status_t psa_load_builtin_key_into_slot( psa_key_slot_t *slot )
+{
+    psa_status_t status = PSA_ERROR_CORRUPTION_DETECTED;
+    psa_key_attributes_t attributes = PSA_KEY_ATTRIBUTES_INIT;
+    psa_key_lifetime_t lifetime = PSA_KEY_LIFETIME_VOLATILE;
+    psa_drv_slot_number_t slot_number = 0;
+    uint8_t *key_buffer = NULL;
+    size_t key_buffer_size = 0;
+    size_t key_buffer_length = 0;
+
+    /* Load keys in the 'builtin' range through their own interface */
+    if( ! psa_key_id_is_builtin(
+            MBEDTLS_SVC_KEY_ID_GET_KEY_ID( slot->attr.id ) ) )
+    {
+        return( PSA_ERROR_DOES_NOT_EXIST );
+    }
+
+    /* Check the platform function to see whether this key actually exists */
+    psa_set_key_id( &attributes, slot->attr.id );
+    status = mbedtls_psa_platform_get_builtin_key(
+                slot->attr.id, &lifetime, &slot_number );
+    if( status != PSA_SUCCESS )
+        return( status );
+
+    /* Set mapped lifetime on the attributes */
+    psa_set_key_lifetime( &attributes, lifetime );
+
+    /* If the key should exist according to the platform, load it through the
+     * driver interface. */
+    status = psa_driver_wrapper_get_key_buffer_size( &attributes,
```

Comment:
This is not addressed and is still relevant I think. I can see it works with the test driver as the size of the buffer does not depend on the key type and size but that's only a specific case. A right sequence seems: call to `psa_driver_wrapper_get_builtin_key` with no buffer to get the key attributes, check for the returned error code to be PSA_ERROR_BUFFER_TOO_SMALL, call `psa_driver_wrapper_get_key_buffer_size`, allocate the buffer and then call again `psa_driver_wrapper_get_builtin_key`.

Reply:
Ah, okay. Now I see what you mean. Yes, let's do that sequence instead.

Reply:
Actually, wouldn't it make more sense to make the driver's `get_builtin_key` function return the expected length through the `key_buffer_length` output when it returns `BUFFER_TOO_SMALL`? That way seems cleaner to me, since there's not really a way to indicate through the key attributes alone that a key is built-in (i.e. the implementation of `get_key_buffer_size` is difficult to rely on when it cannot reliably distinguish whether you're asking for the buffer size of a builtin key or a 'regular' opaque key of the same algorithm, which might be two different sizes, as is the case in the test driver).

I'll make the relevant change to `psa-driver-interface.md` and then you and @gilles-peskine-arm can decide whether you agree.

Reply:
This has been proposed as part of the review of #3878 but decided no to go that way: see [here](https://github.com/ARMmbed/mbedtls/pull/3878#discussion_r561853623).

Reply:
reworked with your proposed flow :)